doc/specs/#1571 - New Tab Menu Customization/#1571 - New Tab Menu Customization.md

```diff
@@ -0,0 +1,125 @@
+the appearance of the new tab dropdown. Let's take a look at an example:
+
+```json
+{
+    "profiles":{ ... },
+    "newTabMenu": [
+        { "type":"profile", "profile": "cmd" },
+        { "type":"profile", "profile": "Windows PowerShell" },
+        { "type":"separator" },
+        {
+            "type":"folder",
+            "name": "ssh",
+            "entries":[
+                { "type":"profile", "profile": "Host 1" },
+                { "type":"profile", "profile": "8.8.8.8" },
+                { "type":"profile", "profile": "Host 2" }
+            ]
+        },
+        { "type":"separator" },
+        { "type":"profile", "profile": "Ubuntu-18.04" },
+        { "type":"profile", "profile": "Fedora" }
+    ]
+}
+```
+
+If a user were to use this as their new tab menu, that they would get is a menu
+that looks like this:
+
+![fig 1](Menu-Customization-000.png)
+
+_fig 1_: A _very rough_ mockup of what this feature might look like
+
+There are three `type`s of objects in this menu:
+* `"type":"profile"`: This is a profile. Clicking on this entry will open a new
+  tab, with that profile. The profile is identified with the `"profile"`
+  parameter, which accepts either a profile `name` or GUID.
+* `"type":"separator"`: This represents a XAML `MenuFlyoutSeparator`, enabling
+  the user to visually space out entries.
+* `"type":"entries"`: This represents a nested menu of entries.
+  - The `"name"` property provides a string of text to display for the group.
+  - The `"entries"` property specifies a list of menu entries that will appear
+    nested under this entry. This can contain other `"type":"entries"` groups as
+    well!
+
+## UI/UX Design
+
+See the above _figure 1_.
+
+The profile's `icon` will also appear as the icon on `profile` entries. If
+there's a keybinding bound to open a new tab with that profile, then that will
+also be added to the `MenuFlyoutItem` as the accelerator text, similar to the
+text we have nowadays.
+
+Beneath the list of profiles will _always_ be the same "Settings", "Feedback"
+and "About" entries, separated by a `MenuFlyoutSeparator`. This is consistent
+with the UI as it exists with no customization. These entries cannot be removed
+with this feature, only the list of profiles customized.
+
+## Capabilities
+
+### Accessibility
+
+This menu will be added to the XAML tree in the same fashion as the current new
+tab flyout, so there should be no dramatic change here.
+
+### Security
+
+_(no change expected)_
+
+### Reliability
+
+_(no change expected)_
+
+### Compatibility
+
+_(no change expected)_
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+I want to make sure that the "new tab with profile index N" shortcut actions
+still use the profiles _in the order specified in the user's settings_, not the
+order they appear in this dropdown. Since a user might have nested profiles in
+this dropdown, picking either a depth-first or breadth-first ordering would be
+confusing.

```

Comment:
What if we have the `index` keybinding arg map to the _space_ of the new tab dropdown?

<kbd>ctrl+shift+1</kbd> --> index: 1 --> 1st MenuFlyoutItem
<kbd>ctrl+shift+2</kbd> --> index: 2 --> 2nd MenuFlyoutItem
etc...

If the targeted MenuFlyoutItem is of...
- `type: profile` --> open the profile in a new tab
- `type: folder` --> open the flyout, and the nested flyout, then the user can choose which profile to open in the submenu

Now here's where things get interesting: how can a user access a nested profile?
Answer: key chords

<kbd>ctrl+shift+1</kbd> + <kbd>ctrl+shift+2</kbd> --> index: 1, subIndex: 2 --> 1st MenuFlyoutItem, 2nd nested MenuFlyoutItem
<kbd>ctrl+shift+2</kbd> + <kbd>ctrl+shift+1</kbd> --> index: 2, subIndex: 1 --> 2nd MenuFlyoutItem, 1st nested MenuFlyoutItem
etc...

We would have to introduce a `subIndex` keybinding arg to get this to work.


In my mind, "index" maps to where it is on the drop down, whereas "profile" (or was it "name"?) maps to the actual profile name. So, this wouldn't break people and might actually be kind of intuitive?

Reply:
I don't hate this a _ton_, but it more changes the meaning of that action to "Open a new tab with the _entry at index N in the new tab dropdown_". Just a bit of a semantic change. Since people would already be opting into this behavior (by configuring this menu), I guess that's okay? 

I think I'd much rather not introduce the `subIndex` param - what if there's lots of nesting?

I think I'd be more okay with just focusing the first element in the sub menu, so the user could navigate it with the keyboard. So for
```
new tab button
├─ Folder 1
│  └─ Profile A
│  └─ Profile B
├─ Folder 2
│  └─ Profile C
│  └─ Profile D
└─ Profile E
```

* <kbd>ctrl+shift+1</kbd> focuses "Profile A", but the user needs to press enter/space to launch it
* <kbd>ctrl+shift+2</kbd> focuses "Profile C", but the user needs to press enter/space to launch it
* <kbd>ctrl+shift+3</kbd> launches the profile in Profile E

I'd definitely want more feedback on this particular point though.

Reply:
> I don't hate this a _ton_, but it more changes the meaning of that action to "Open a new tab with the _entry at index N in the new tab dropdown_". Just a bit of a semantic change. Since people would already be opting into this behavior (by configuring this menu), I guess that's okay?

Still a semantic change, but I wanted to point out that this is what we say in the schema:
> "Profile that will open based on its position in the dropdown (starting at 0)."
So I'm ok with it :P haha

> I think I'd much rather not introduce the `subIndex` param - what if there's lots of nesting?

Whatcha mean here? You mean something like this:
```
new tab button
├─ Folder 1
│  └─ Folder 1.a
│     └─ Profile A
│     └─ Profile B
│  └─ Folder 1.b
│     └─ Profile C
│     └─ Profile D
```
Is that allowed? If so, then "subIndex" should just do the behavior you mentioned about focusing on the relevant part of the flyout menu.

> I think I'd be more okay with just focusing the first element in the sub menu, so the user could navigate it with the keyboard. So for
> 
> ```
> new tab button
> ├─ Folder 1
> │  └─ Profile A
> │  └─ Profile B
> ├─ Folder 2
> │  └─ Profile C
> │  └─ Profile D
> └─ Profile E
> ```
> 
> * ctrl+shift+1 focuses "Profile A", but the user needs to press enter/space to launch it
> * ctrl+shift+2 focuses "Profile C", but the user needs to press enter/space to launch it
> * ctrl+shift+3 launches the profile in Profile E
> 
> I'd definitely want more feedback on this particular point though.

I'd be ok with this functionality either way, at least for v1 of this feature.
I guess if people want subindex, the workaround is to link the profile name to the keybinding instead of index. And that's not a _bad_ workaround.

Reply:
> Whatcha mean here? You mean something like this:

Yea, that's exactly the scenario I'm thinking of. 

>  I guess if people want subindex, the workaround is to link the profile name to the keybinding instead of index. And that's not a bad workaround.

Oh, that's definitely the right solution here - might as well not introduce another unnecessary feature layer here when `profile` will work just fine.

I'd love to get an ACK from another person on the team on this particular point before I go change this section in the spec though 😉

Reply:
Oh no. Now that I'm really thinking that the `{ "type": "action", "action": {...} }` "Future Consideration" should really be a main part of this, I'm more worried about this.

The `index` param works for both `newTab` _and_ `splitPane`. What do we expect to happen when someone executes "split pane with entry 3", and entry 3 is a nested action (as described in #5970) that's basically "open a new tab with foo, split that with bar"? 

What happens when someone says "new tab with entry 2", and entry 2 is "copy"? That doesn't make any sense!

I guess for that scenario we could just make `"index": <index of an action entry>` just no-op. Then, we'd validate that all `newTab`/`splitPane` actions with indicies point either to a `folder` or `profile` entry, and display a warning if the user sets that. 

Okay I talked myself back into it.

Reply:
@leonMSFT I agree with that. I think I added a section on the `openNewTabDropdown` action to try and to the "new tab menu" order version of "open new tab".

> To cover that above scenario, we could consider adding an `index` parameter to
> the `openNewTabDropdown` action. If specified, that would open either the N'th
> action in the dropdown (ignoring seperators), or open the dropdown with the n'th
> item selected.

I definitely concur that we should be shying away from the "index" arg nowadays